File: src/api/identity.ts
import type { HttpClient } from "./http.ts";

interface MeResponse {
  response: {
    id: number;
    business_memberships: Array<{
      business: {
        id: number;
        account_id: string;
      };
    }>;
  };
}

export interface Identity {
  account_id: string;
  business_id: number;
}

export async function getIdentity(http: HttpClient): Promise<Identity> {
  const data = await http.get<MeResponse>("/auth/api/v1/users/me");
  const memberships = data.response.business_memberships;

  if (memberships.length === 0) {
    throw new Error("No business memberships found on this account.");
  }

  const first = memberships[0]!;
  return {
    account_id: first.business.account_id,
    business_id: first.business.id,
  };
}

File: internal/api/identity.go
package api

import "fmt"

// Identity holds the account and business IDs for a FreshBooks user.
type Identity struct {
	AccountID  string
	BusinessID int
}

type meResponse struct {
	Response struct {
		ID                  int `json:"id"`
		BusinessMemberships []struct {
			Business struct {
				ID        int    `json:"id"`
				AccountID string `json:"account_id"`
			} `json:"business"`
		} `json:"business_memberships"`
	} `json:"response"`
}

// GetIdentity fetches the current user's identity from the FreshBooks API.
func GetIdentity(c *HttpClient) (*Identity, error) {
	var data meResponse
	if err := c.Get("/auth/api/v1/users/me", nil, &data); err != nil {
		return nil, err
	}

	memberships := data.Response.BusinessMemberships
	if len(memberships) == 0 {
		return nil, fmt.Errorf("no business memberships found on this account")
	}

	first := memberships[0]
	return &Identity{
		AccountID:  first.Business.AccountID,
		BusinessID: first.Business.ID,
	}, nil
}
